command/agent/cache/lease_cache.go

```diff
@@ -0,0 +1,743 @@
+		return errInvalidType
+	}
+
+	c.logger.Debug("successfully cleared matching cache entries")
+
+	return nil
+}
+
+// handleRevocationRequest checks whether the originating request is a
+// revocation request, and if so perform applicable cache cleanups.
+// Returns true is this is a revocation request.
+func (c *LeaseCache) handleRevocationRequest(ctx context.Context, req *SendRequest, resp *SendResponse) (bool, error) {
+	// Lease and token revocations return 204's on success. Fast-path if that's
+	// not the case.
+	if resp.Response.StatusCode != http.StatusNoContent {
+		return false, nil
+	}
+
+	namespace, path := deriveNamespaceAndRevocationPath(req)
+
+	switch {
+	case path == vaultPathTokenRevoke:
+		// Get the token from the request body
+		jsonBody := map[string]interface{}{}
+		if err := json.Unmarshal(req.RequestBody, &jsonBody); err != nil {
+			return false, err
+		}
+		token, ok := jsonBody["token"]
+		if !ok {
+			return false, fmt.Errorf("failed to get token from request body")
+		}
+
+		// Clear the cache entry associated with the token and all the other
+		// entries belonging to the leases derived from this token.
+		if err := c.handleCacheClear(ctx, "token", token.(string)); err != nil {
+			return false, err
+		}
+
+	case path == vaultPathTokenRevokeSelf:
+		// Clear the cache entry associated with the token and all the other
+		// entries belonging to the leases derived from this token.
+		if err := c.handleCacheClear(ctx, "token", req.Token); err != nil {
+			return false, err
+		}
+
+	case path == vaultPathTokenRevokeAccessor:
+		jsonBody := map[string]interface{}{}
+		if err := json.Unmarshal(req.RequestBody, &jsonBody); err != nil {
+			return false, err
+		}
+		accessor, ok := jsonBody["accessor"]
+		if !ok {
+			return false, fmt.Errorf("failed to get accessor from request body")
+		}
+
+		if err := c.handleCacheClear(ctx, "token_accessor", accessor.(string)); err != nil {
+			return false, err
+		}
+
+	case path == vaultPathTokenRevokeOrphan:
+		jsonBody := map[string]interface{}{}
+		if err := json.Unmarshal(req.RequestBody, &jsonBody); err != nil {
+			return false, err
+		}
+		token, ok := jsonBody["token"]
+		if !ok {
+			return false, fmt.Errorf("failed to get token from request body")
+		}
+
+		// Find out all the indexes that are directly tied to the revoked token
+		indexes, err := c.db.GetByPrefix(cachememdb.IndexNameToken.String(), token.(string))
+		if err != nil {
+			return false, err
+		}
+
+		// Out of these indexes, one will be for the token itself and the rest
+		// will be for leases of this token. Cancel the contexts of all the
+		// leases and return from renewer goroutine for the token's index
+		// without cancelling the context. Cancelling the context of the
+		// token's renewer will evict all the child tokens which is not
+		// desired.
+		for _, index := range indexes {
+			if index.Lease != "" {
+				index.RenewCtxInfo.CancelFunc()
+			} else {
+				close(index.RenewCtxInfo.DoneCh)
```

Comment:
Why are we closing this? Wouldn't this evict all child tokens from cache?

Reply:
This is not cancelling the context and hence the child tokens will not be evicted. It is closing the channel that returns the renewer goroutine. So, it only removes a single item.

Reply:
But which item are we removing? we don't want to stop renewing child tokens, right?

Reply:
If an intermediate token is revoke, its parent and parents lease should be intact. The child tokens and their respective leases should be intact. Only the revoked token and the leases belonging to the revoked token should be gone. I'll definitely need a test for this case which I am working on, but the code seems to be doing the right thing. The child tokens of the tokens being revoked are being manually walked over and cancelled individually, and then the renewer of the revoked token is inform to return without cancelling its context. That way the children contexts are unaffected.

Reply:
Added a test to cover this case. It's working as expected.